packages/medusa/src/services/totals-new.ts

```diff
@@ -0,0 +1,692 @@
+      }
+    }
+
+    const calculationMethod = taxRate
+      ? this.getShippingMethodTotalsLegacy.bind(this)
+      : this.getShippingMethodTotals.bind(this)
+
+    const shippingMethodsTotals: {
+      [lineItemId: string]: ShippingMethodTotals
+    } = {}
+    for (const shippingMethod of shippingMethods) {
+      shippingMethodsTotals[shippingMethod.id] = await calculationMethod(
+        shippingMethod,
+        {
+          includeTax,
+          calculationContext,
+          taxLines: shippingMethodsTaxLinesMap[shippingMethod.id],
+          discounts,
+        }
+      )
+    }
+
+    return shippingMethodsTotals
+  }
+
+  /**
+   * Calcul and return the shipping method totals
+   * @param shippingMethod
+   * @param includeTax
+   * @param calculationContext
+   * @param taxLines
+   * @param discounts
+   */
+  async getShippingMethodTotals(
+    shippingMethod: ShippingMethod,
+    {
+      includeTax,
+      calculationContext,
+      taxLines,
+      discounts,
+    }: {
+      includeTax?: boolean
+      calculationContext: TaxCalculationContext
+      taxLines?: (ShippingMethodTaxLine | LineItemTaxLine)[]
+      discounts?: Discount[]
+    }
+  ) {
+    const totals: ShippingMethodTotals = {
+      price: shippingMethod.price,
+      original_total: shippingMethod.price,
+      total: shippingMethod.price,
+      subtotal: shippingMethod.price,
+      original_tax_total: 0,
+      tax_total: 0,
+      tax_lines: (taxLines ??
+        shippingMethod.tax_lines ??
+        []) as ShippingMethodTaxLine[],
+    }
+
+    const calculationContext_: TaxCalculationContext = {
+      ...calculationContext,
+      shipping_methods: [shippingMethod],
+    }
+
+    // Force the tax lines to exist anyway
+    if (includeTax && !totals.tax_lines.length) {
+      throw new MedusaError(
+        MedusaError.Types.UNEXPECTED_STATE,
+        "Tax Lines must be joined to calculate taxes"
+      )
+    }
+
+    const includesTax =
+      this.featureFlagRouter_.isFeatureEnabled(
+        TaxInclusivePricingFeatureFlag.key
+      ) && shippingMethod.includes_tax
+
+    totals.original_tax_total = await this.taxCalculationStrategy_.calculate(
+      [],
+      totals.tax_lines,
+      calculationContext_
+    )
+    totals.tax_total = totals.original_tax_total
+
+    if (includesTax) {
+      totals.subtotal -= totals.tax_total
+    } else {
+      totals.original_total += totals.original_tax_total
+      totals.total += totals.tax_total
+    }
+
+    const hasFreeShipping = discounts?.some(
+      (d) => d.rule.type === DiscountRuleType.FREE_SHIPPING
+    )
+
+    if (hasFreeShipping) {
+      totals.total = 0
+      totals.subtotal = 0
+      totals.tax_total = 0
+    }
+
+    return totals
+  }
+
+  /**
+   * Calcul and return the shipping method totals legacy using teh tax rate
+   * @param shippingMethod
+   * @param calculationContext
+   * @param taxLines
+   * @param discounts
+   */
+  async getShippingMethodTotalsLegacy(
```

Comment:
can we delete?

Reply:
probably not, related to https://github.com/medusajs/medusa/pull/2475#discussion_r1012768714